src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -326,31 +401,28 @@ private FsTranslogFile translogForLocation(Location location) {
      */
     @Override
     public Translog.Operation read(Location location) {
-        rwl.readLock().lock();
-        try {
-            FsTranslogFile translog = translogForLocation(location);
-            if (translog != null) {
-                byte[] data = translog.read(location);
-                try (BytesStreamInput in = new BytesStreamInput(data)) {
-                    // Return the Operation using the current version of the
-                    // stream based on which translog is being read
-                    return translog.getStream().read(in);
+        try (ReleasableLock _ = readLock.acquire()) {
+            FsChannelReader reader = null;
+            if (current.translogId() == location.translogId) {
+                reader = current;
+            } else {
+                for (FsChannelReader translog : uncomittedTranslogs) {
```

Comment:
Just a random notion, but it would be nice if we had a datastructure for doing lookups by translogid, maybe we could use a `LinkedHashMap` to preserve insertion order but still allow lookup-by-id?